src/Agent/NewRelic/Agent/Core/DataTransport/Client/NRWebRequestClient.cs

```diff
@@ -0,0 +1,98 @@
+        {
+            _configuration = configuration;
+        }
+
+        public override async Task<IHttpResponse> SendAsync(IHttpRequest request)
+        {
+            try
+            {
+                _httpWebRequest = _getHttpWebRequestFunc(request.Uri);
+
+                // If a null assignment is made it will bypass the default (IE) proxy settings 
+                // bypassing those settings could cause 504s to be thrown where the user has 
+                // implemented a proxy via IE instead of implementing an external proxy and declaring the values in the New Relic config
+                if (_proxy != null)
+                {
+                    _httpWebRequest.Proxy = _proxy;
+                }
+
+                _httpWebRequest.KeepAlive = true;
+                _httpWebRequest.Timeout = (int)request.Timeout.TotalMilliseconds;
+                _httpWebRequest.ContentType = request.Content.ContentType;
+                _httpWebRequest.UserAgent = $"NewRelic-DotNetAgent/{AgentInstallConfiguration.AgentVersion}";
+                _httpWebRequest.Method = _configuration.PutForDataSend ? "PUT" : "POST";
+                _httpWebRequest.ContentLength = request.Content.PayloadBytes.Length;
+
+                _httpWebRequest.Headers.Add("ACCEPT-ENCODING", "gzip");
+                _httpWebRequest.Headers.Add("CONTENT-ENCODING", request.Content.Encoding);
+
+                foreach (var header in request.Headers)
+                {
+                    _httpWebRequest.Headers.Add(header.Key, header.Value);
+                }
+
+                using (var outputStream = _httpWebRequest.GetRequestStream())
+                {
+                    if (outputStream == null)
+                    {
+                        throw new NullReferenceException("outputStream");
+                    }
+
+                    // .ConfigureAwait(false) is required here for some reason
+                    await outputStream.WriteAsync(request.Content.PayloadBytes, 0, (int)_httpWebRequest.ContentLength).ConfigureAwait(false);
+                }
+
+                // .ConfigureAwait(false) is required here for some reason
+                var resp = (HttpWebResponse)await _httpWebRequest.GetResponseAsync().ConfigureAwait(false);
+
+                return new WebRequestClientResponse(request.RequestGuid, resp);
+            }
+            catch (WebException ex)
+            {
+                if (ex.Response is HttpWebResponse httpWebResponse)
+                    return new WebRequestClientResponse(request.RequestGuid, httpWebResponse);
+
+                if (_diagnoseConnectionError)
+                {
+                    DiagnoseConnectionError(request.Uri.Host);
+                }
+
+                throw;
+            }
+        }
+
+        public override void Dispose()
+        {
+        }
+
+        // for unit testing only
+        public void SetHttpWebRequestFunc(Func<Uri, HttpWebRequest> fooFunc )
```

Comment:
We should probably rename `fooFunc` to something else.

Reply:
Alternatively you might be able to avoid needing this method if you just passed the create method as a .ctor parameter. That method can be registered with the IOC container, and the test can define its own method.

Reply:
Gonna leave the method in place for now.